rustdoc: render generic params & where-clauses of cross-crate assoc tys in impls

fmease opened the pull request:
We used to only ever render generic parameters & where-clauses of cross-crate associated types when the item was located inside of a trait and we used to just drop them when it was inside of an impl block (trait or inherent).

Fixes #112904.

@rustbot label A-cross-crate-reexports

fmease committed:
rustdoc: render gen params & where-clauses of cross-crate assoc tys in impl blocks

bors commented:
:umbrella: The latest upstream changes (presumably #112974) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease commented:
Note to reviewers: The diff looks unnecessarily complex (split diff is slightly better). Basically I just moved the code responsible for obtaining the `generics` and the `predicates` (for the where-clause) out of the `TraitContainer` branch as alluded to in the PR description allowing us to access them in the `ImplContainer` branch.

Apart from that, I added an extra `TraitContainer` branch that makes sure we don't try to obtain `item_bounds` inside an `ImplContainer` which would be incorrect (it'd lead to an ICE if we tried btw). See the test files for more context about that.

(I've also changed the indentation of the if/let-chains found in the code block I've moved since it didn't align with the style guide (`rustfmt` cannot handle them yet) and it started to annoy me. I originally wrote that part of the code. If you are wondering, it doesn't make the diff worse)

GuillaumeGomez commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit c643bedf7c90940005da021ba80b6c4e6dcbdaa5 has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez c643bedf7c90940005da021ba80b6c4e6dcbdaa5 -->
<!-- homu: {"type":"Approved","sha":"c643bedf7c90940005da021ba80b6c4e6dcbdaa5","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #113057) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
